In [ ]:
import pandas as pd

In [ ]:
!file WA_Fn-UseC_-Telco-Customer-Churn.csv

WA_Fn-UseC_-Telco-Customer-Churn.csv: CSV text


In [ ]:
# Read CSV File

df_loaded = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")

df_loaded

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes


In [ ]:
import pandas as pd

In [ ]:
!file WA_Fn-UseC_-Telco-Customer-Churn.csv

WA_Fn-UseC_-Telco-Customer-Churn.csv: CSV text


In [ ]:
import pandas as pd

df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv', encoding='latin1')

In [ ]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 7043
Columns: 21


In [ ]:
print(df.columns.tolist())

['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn']


In [ ]:
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
Partner,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


In [ ]:
#Convert TotalCharges to numeric:

df['TotalCharges'] = pd.to_numeric(
    df['TotalCharges'],
    errors='coerce'
)

In [ ]:
df['TotalCharges'].fillna(
    df['TotalCharges'].median(),
    inplace=True
)

/tmp/ipykernel_1816/1685388760.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['TotalCharges'].fillna(


In [ ]:
#Calculate Churn Rate

churn_rate = (
    df['Churn']
      .value_counts(normalize=True) * 100
)

print(churn_rate)

Churn
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64


In [ ]:
#Exploratory Analysis
#Churn by Contract

pd.crosstab(
    df['Contract'],
    df['Churn']
)

Churn,No,Yes
Contract,,
Month-to-month,2220,1655
One year,1307,166
Two year,1647,48


In [ ]:
#Average Monthly Charges

df.groupby('Churn')[
    'MonthlyCharges'
].mean()

,MonthlyCharges
Churn,
No,61.265124
Yes,74.441332


In [ ]:
#Average Tenure

df.groupby('Churn')[
    'tenure'
].mean()

,tenure
Churn,
No,37.569965
Yes,17.979133


In [ ]:
#Prepare for SQL Server

#Save cleaned data:

df.to_csv(
    'cleaned_customer_churn.csv',
    index=False
)

In [ ]:
df.shape
df.columns.tolist()
df['Churn'].value_counts()

,count
Churn,
No,5174
Yes,1869


In [ ]:
#Download the CSV from Google Colab

from google.colab import files
files.download('cleaned_customer_churn.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# Customer Churn Analysis Project

**Author:** Atiku Abubakar
**Tools:** Python, SQL Server, Power BI, Google Colab

---

# 1. Business Problem

A telecommunications company wants to understand why customers leave (churn), identify high-risk customers, and develop strategies to improve customer retention.

---

# 2. Project Objectives

* Calculate the customer churn rate.
* Identify the characteristics of customers who churn.
* Analyze the impact of contract type, tenure, and monthly charges on churn.
* Build an interactive dashboard and provide business recommendations.

---

# 3. Import Libraries

```python
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
```

---

# 4. Load the Dataset

```python
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')
df.head()
```

---

# 5. Data Understanding

```python
df.shape
df.info()
df.describe()
df.isnull().sum()
```

## Findings

* Total Customers: 7,043
* Dataset contains customer demographics, services, contracts, and churn information.
* Missing values were found in the `TotalCharges` column.

---

# 6. Data Cleaning

```python
df['TotalCharges'] = pd.to_numeric(
    df['TotalCharges'],
    errors='coerce'
)

df['TotalCharges'].fillna(
    df['TotalCharges'].median(),
    inplace=True
)
```

## Cleaning Summary

* Converted `TotalCharges` to numeric.
* Replaced missing values with the median.
* Verified that there were no remaining missing values.

---

# 7. Exploratory Data Analysis (EDA)

## Customer Churn Distribution

```python
df['Churn'].value_counts()
```

### Results

| Churn | Customers |
| ----- | --------: |
| No    |      5174 |
| Yes   |      1869 |

---

## Churn Rate

```python
df['Churn'].value_counts(normalize=True) * 100
```

### Results

| Churn | Percentage |
| ----- | ---------: |
| No    |     73.47% |
| Yes   |     26.53% |

---

## Contract Type Analysis

```python
pd.crosstab(df['Contract'], df['Churn'])
```

---

## Monthly Charges Analysis

```python
df.groupby('Churn')['MonthlyCharges'].mean()
```

---

## Tenure Analysis

```python
df.groupby('Churn')['tenure'].mean()
```

---

# 8. Visualizations

## Churn Distribution

```python
df['Churn'].value_counts().plot(kind='bar')
plt.title('Customer Churn Distribution')
plt.show()
```

---

# 9. Key Insights

1. Approximately 26.53% of customers churned.
2. Customers on month-to-month contracts churn more frequently.
3. Customers with higher monthly charges tend to churn more.
4. Customers with shorter tenure are at higher risk of churn.

---

# 10. Business Recommendations

1. Introduce loyalty and retention programs.
2. Encourage customers to move to longer-term contracts.
3. Target high-risk customers with personalized offers.
4. Improve customer service and support experience.

---

# 11. Export Cleaned Dataset

```python
df.to_csv(
    'cleaned_customer_churn.csv',
    index=False
)
```

---

# Project Conclusion

This analysis identified the main drivers of customer churn and provided actionable recommendations to improve customer retention and reduce revenue loss.

**Next Steps:**

* Perform SQL analysis.
* Build a Power BI dashboard.
* Develop a predictive churn model.
